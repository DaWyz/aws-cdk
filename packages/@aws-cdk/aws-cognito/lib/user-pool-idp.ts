import { Construct, IResource, Resource } from '@aws-cdk/core';
import { CfnUserPoolIdentityProvider } from './cognito.generated';
import { IUserPool } from './user-pool';

/**
 * Represents a UserPoolIdentityProvider
 */
export interface IUserPoolIdentityProvider extends IResource {
  /**
   * The primary identifier of this identity provider
   * @attribute
   */
  readonly providerName: string;
}

/**
 * The properties to initialize a new UserPoolIdentityProvider
 */
export interface UserPoolIdentityProviderProps {
  /**
   * The name of this provider. This will be its primary identifier.
   */
  readonly userPoolIdentityProviderName: string;

  /**
   * Options to integrate with third party social identity providers such as Facebook, Google, Amazon and Apple.
   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-identity-provider.html#cognito-user-pools-facebook-provider
   */
  readonly socialIdentity: SocialIdentityProvider;

  /**
   * The user pool to whom this provider is attached to.
   */
  readonly userPool: IUserPool;
}

/**
 * Options to integrate with the various social identity providers.
 */
export class SocialIdentityProvider {
  /**
   * Federate with 'Facebook Login'
   * @see https://developers.facebook.com/docs/facebook-login/
   */
  public static facebook(options: FacebookProviderOptions) {
    const scopes = options.scopes ?? [ 'public_profile' ];
    return new SocialIdentityProvider('Facebook', {
      client_id: options.clientId,
      client_secret: options.clientSecret,
      authorize_scopes: scopes.join(','),
      api_version: options.apiVersion,
    });
  }

  /**
   * Federate with 'Google Sign In'
   * @see https://developers.google.com/identity/
   */
  public static google(options: GoogleProviderOptions) {
    const scopes = options.scopes ?? [ 'profile', 'email', 'openid' ];
    return new SocialIdentityProvider('Google', {
      client_id: options.clientId,
      client_secret: options.clientSecret,
      authorize_scopes: scopes.join(' '),
    });
  }

  /**
   * Federate with 'Login with Amazon'
   * @see https://developer.amazon.com/apps-and-games/login-with-amazon
   */
  public static amazon(options: AmazonProviderOptions) {
    const scopes = options.scopes ?? [ 'profile' ];
    return new SocialIdentityProvider('LoginWithAmazon', {
      client_id: options.clientId,
      client_secret: options.clientSecret,
      authorize_scopes: scopes.join(' '),
    });
  }

  /**
   * Federate with 'Sign in with Apple'
   * @see https://developer.apple.com/sign-in-with-apple/
   */
  public static apple(options: AppleProviderOptions) {
    const scopes = options.scopes ?? [ 'public_profile', 'email' ];
    return new SocialIdentityProvider('SignInWithApple', {
      client_id: options.servicesId,
      team_id: options.teamId,
      key_id: options.keyId,
      private_key: options.privateKey,
      authorize_scopes: scopes.join(' '),
    });
  }

  /**
   * Custom configuration that is not yet supported by the CDK.
   */
  public static custom(providerType: string, providerDetails: { [key: string]: any }) {
    return new SocialIdentityProvider(providerType, providerDetails);
  }

  // tslint:disable:max-line-length
  /**
   * The type of the provider as recognized by CloudFormation
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolidentityprovider.html#cfn-cognito-userpoolidentityprovider-providertype
   */
  public readonly providerType: string;
  /**
   * The properties needed to connect to the provider as recognized by CloudFormation
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolidentityprovider.html#cfn-cognito-userpoolidentityprovider-providerdetails
   */
  public readonly providerDetails: { [key: string]: any };
  // tslint:enable

  private constructor(providerType: string, providerDetails: { [key: string]: any }) {
    this.providerType = providerType;
    this.providerDetails = providerDetails;
  }
}

/**
 * Options to integrate with 'Facebook Login'.
 */
export interface FacebookProviderOptions {
  /**
   * The client id recognized by Facebook APIs.
   */
  readonly clientId: string;
  /**
   * The client secret to be accompanied with clientUd for Facebook to authenticate the client.
   * @see https://developers.facebook.com/docs/facebook-login/security#appsecret
   */
  readonly clientSecret: string;
  /**
   * The list of facebook permissions to obtain for getting access to the Facebook profile.
   * @see https://developers.facebook.com/docs/facebook-login/permissions
   * @default [ public_profile ]
   */
  readonly scopes?: string[];
  /**
   * The Facebook API version to use
   * @default - to the oldest version supported by Facebook
   */
  readonly apiVersion?: string;
}

/**
 * Options to integrate with 'Google Sign in'.
 */
export interface GoogleProviderOptions {
  /**
   * The client id recognized by 'Google Sign in'.
   */
  readonly clientId: string;
  /**
   * The client secret to be accompanied with clientId for Google to authenticate the client.
   */
  readonly clientSecret: string;
  /**
   * The list of Google permissions to obtain for getting access to the Google profile.
   * @see https://developers.google.com/identity/protocols/oauth2/scopes
   * @default [ profile, email, openid ]
   */
  readonly scopes?: string[];
}

/**
 * Options to integrate with 'Login with Amazon'.
 */
export interface AmazonProviderOptions {
  /**
   * The client id recognized by 'Login with Amazon' APIs.
   * @see https://developer.amazon.com/docs/login-with-amazon/security-profile.html#client-identifier
   */
  readonly clientId: string;
  /**
   * The client secret to be accompanied with clientId for 'Login with Amazon' APIs to authenticate the client.
   * @see https://developer.amazon.com/docs/login-with-amazon/security-profile.html#client-identifier
   */
  readonly clientSecret: string;
  /**
   * The types of user profile data to obtain for the Amazon profile.
   * @see https://developer.amazon.com/docs/login-with-amazon/customer-profile.html
   * @default [ profile ]
   */
  readonly scopes?: string[];
}

/**
 * Options to integrate with 'Sign in with Apple'.
 */
export interface AppleProviderOptions {
  /**
   * The Services id received when the 'Sign in with Apple' client was created.
   */
  readonly servicesId: string;
  /**
   * The team id received when the 'Sign in with Apple' client was created.
   */
  readonly teamId: string;
  /**
   * The key id received when the 'Sign in with Apple' client was created.
   */
  readonly keyId: string;
  /**
   * The private key received when the 'Sign in with Apple' client was created.
   */
  readonly privateKey: string;
  /**
   * The types of user profile data to obtain for the Amazon profile.
   * @see https://developer.amazon.com/docs/login-with-amazon/customer-profile.html
   * @default [ public_profile, email ]
   */
  readonly scopes?: string[];
}

/**
 * Define a identity provider for a user pool.
 * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-identity-provider.html
 */
export class UserPoolIdentityProvider extends Resource implements IUserPoolIdentityProvider {
  public readonly providerName: string;

  constructor(scope: Construct, id: string, props: UserPoolIdentityProviderProps) {
    super(scope, id, {
      physicalName: props.userPoolIdentityProviderName,
    });

    const resource = new CfnUserPoolIdentityProvider(this, 'Resource', {
      providerName: this.physicalName,
      userPoolId: props.userPool.userPoolId,
      providerType: props.socialIdentity.providerType,
      providerDetails: props.socialIdentity.providerDetails,
    });
    this.providerName = super.getResourceNameAttribute(resource.ref);
  }
}