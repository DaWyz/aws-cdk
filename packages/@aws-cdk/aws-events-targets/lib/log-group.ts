import * as events from '@aws-cdk/aws-events';
import * as logs from '@aws-cdk/aws-logs';
/**
 * Customize the CloudWatch LogGroup Event Target
 */
export interface LogGroupProps {
  /**
   * The event to send to the CloudWatch LogGroup
   *
   * This will be the event logged into the CloudWatch LogGroup
   *
   * @default - the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;
}

/**
 * Use an AWS CloudWatch LogGroup as an event rule target
 */
export class LogGroup implements events.IRuleTarget {
  constructor(private readonly handler: logs.ILogGroup, private readonly props: LogGroupProps = {}) {}

  /**
   * Returns a RuleTarget that can be used to log an event into a CloudWatch LogGroup
   */
  public bind(_rule: events.IRule, _id?: string): events.RuleTargetConfig {
    return {
      id: '',
      arn: this.handler.logGroupArn,
      input: this.props.event,
      targetResource: this.handler,
    };
  }
}
