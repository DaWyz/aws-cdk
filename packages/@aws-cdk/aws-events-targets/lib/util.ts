import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as sqs from '@aws-cdk/aws-sqs';
import { Construct, ConstructNode, IConstruct, Names, Stack } from '@aws-cdk/core';

/**
 * Obtain the Role for the EventBridge event
 *
 * If a role already exists, it will be returned. This ensures that if multiple
 * events have the same target, they will share a role.
 */
export function singletonEventRole(scope: IConstruct, policyStatements: iam.PolicyStatement[]): iam.IRole {
  const id = 'EventsRole';
  const existing = scope.node.tryFindChild(id) as iam.IRole;
  if (existing) { return existing; }

  const role = new iam.Role(scope as Construct, id, {
    assumedBy: new iam.ServicePrincipal('events.amazonaws.com'),
  });

  policyStatements.forEach(role.addToPolicy.bind(role));

  return role;
}

/**
 * Allows a Lambda function to be called from a rule
 */
export function addLambdaPermission(rule: events.IRule, handler: lambda.IFunction): void {
  let scope: Construct | undefined;
  let node: ConstructNode = handler.permissionsNode;
  if (rule instanceof Construct) {
    // Place the Permission resource in the same stack as Rule rather than the Function
    // This is to reduce circular dependency when the lambda handler and the rule are across stacks.
    scope = rule;
    node = rule.node;
  }
  const permissionId = `AllowEventRule${Names.nodeUniqueId(rule.node)}`;
  if (!node.tryFindChild(permissionId)) {
    handler.addPermission(permissionId, {
      scope,
      action: 'lambda:InvokeFunction',
      principal: new iam.ServicePrincipal('events.amazonaws.com'),
      sourceArn: rule.ruleArn,
    });
  }
}


export function addToDeadLetterQueueResourcePolicy(rule: events.IRule, queue: sqs.IQueue) {
  const ruleParsedStack = Stack.of(rule);
  const queueParsedStack = Stack.of(queue);

  if (ruleParsedStack.region !== queueParsedStack.region) {
    throw new Error(`Cannot assign Dead Letter Queue to the rule ${rule}. Both the queue and the rule must be in the same region`);
  }

  // Skip Resource Policy creation if the Queue is not in the same account.
  // There is no way to add a target onto an imported rule, so we can assume we will run the following code only
  // in the account where the rule is created.

  if (ruleParsedStack.account === queueParsedStack.account) {
    const policyStatementId = `AllowEventRule${Names.nodeUniqueId(rule.node)}`;

    queue.addToResourcePolicy(new iam.PolicyStatement({
      sid: policyStatementId,
      principals: [new iam.ServicePrincipal('events.amazonaws.com')],
      effect: iam.Effect.ALLOW,
      actions: ['sqs:SendMessage'],
      resources: [queue.queueArn],
      conditions: {
        ArnEquals: {
          'aws:SourceArn': rule.ruleArn,
        },
      },
    }));
  } else {
    // Maybe we could post a warning telling the user to create the permission in the target account manually ?
  }
}
