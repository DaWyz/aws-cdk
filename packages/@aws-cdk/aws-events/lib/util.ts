import { EventFilter } from './event-filter';

/**
 * Merge the `src` event pattern into the `dest` event pattern by adding all
 * values from `src` into the fields in `dest`.
 *
 * See `rule.addEventPattern` for details.
 */
export function mergeEventPattern(dest: any, src: any) {
  dest = dest || { };

  mergeObject(formatObject(dest), formatObject(src));

  return dest;

  function transformValue(value: EventFilter | any) {
    if (value instanceof EventFilter) {
      return value.conditions;
    }

    return value;
  }

  function formatObject(obj: any) {
    for (const [field, value] of Object.entries(obj)) {
      if (value instanceof EventFilter) {
        obj[field] = value.conditions;
        continue;
      }

      if (Array.isArray(value)) {
        let filters = [];
        for (let filter of value) {
          filters.push(transformValue(filter));
        }

        obj[field] = value;
        continue;
      }

      if (typeof(value) === 'object') {
        obj[field] = formatObject(value);
      }
    }

    return obj;
  }

  function mergeObject(destObj: any, srcObj: any) {
    if (typeof(srcObj) !== 'object') {
      throw new Error(`Invalid event pattern '${JSON.stringify(srcObj)}', expecting an object or an array`);
    }

    for (const field of Object.keys(srcObj)) {
      const srcValue = srcObj[field];
      const destValue = destObj[field];

      if (srcValue === undefined) { continue; }

      if (typeof(srcValue) !== 'object') {
        throw new Error(`Invalid event pattern field { ${field}: ${JSON.stringify(srcValue)} }. All fields must be arrays`);
      }

      // dest doesn't have this field
      if (destObj[field] === undefined) {
        destObj[field] = srcValue;
        continue;
      }

      if (Array.isArray(srcValue) !== Array.isArray(destValue)) {
        throw new Error(`Invalid event pattern field ${field}. ` +
          `Type mismatch between existing pattern ${JSON.stringify(destValue)} and added pattern ${JSON.stringify(srcValue)}`);
      }

      // if this is an array, concat and deduplicate the values
      if (Array.isArray(srcValue)) {
        const result = [...destValue, ...srcValue];
        const resultJson = result.map(i => JSON.stringify(i));
        destObj[field] = result.filter((value, index) => resultJson.indexOf(JSON.stringify(value)) === index);
        continue;
      }

      // otherwise, it's an object, so recurse
      mergeObject(destObj[field], srcValue);
    }
  }
}
