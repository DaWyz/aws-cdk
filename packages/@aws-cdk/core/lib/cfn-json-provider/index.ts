/**
 * Parses the value of "Value" and reflects it back as attribute.
 */
export async function handler(event: any) {
  return {
    Data: {
      Value: JSON.parse(event.ResourceProperties.Value),
    },
  };
}