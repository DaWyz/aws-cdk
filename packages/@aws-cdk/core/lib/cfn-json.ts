import { Construct } from './construct-compat';
import { CustomResource } from './custom-resource';
import { CustomResourceProvider, CustomResourceProviderRuntime } from './custom-resource-provider';
import { Reference } from './reference';
import { IResolvable, IResolveContext } from './resolvable';
import { Stack } from './stack';
import { captureStackTrace } from './stack-trace';

export interface CfnJsonProps {
  /**
   * The value to resolve. Can be any JavaScript object, including tokens and
   * references in keys or values.
   */
  readonly value: any;
}

/**
 * Captures a synthesis-time JSON object a CloudFormation reference which
 * resolves during deployment to the resolved values of the JSON object.
 *
 * The main use case for this is to overcome a limitation in CloudFormation that
 * does not allow using intrinsic functions as dictionary keys (because
 * dictionary keys in JSON must be strings). Specifically this is common in IAM
 * conditions such as `StringEquals: { lhs: "rhs" }` where you want "lhs" to be
 * a reference.
 *
 * This object is resolvable, so it can be used as a value.
 *
 * This construct is backed by a custom resource.
 */
export class CfnJson extends Construct implements IResolvable {
  public readonly creationStack: string[] = [];

  /**
   * An Fn::GetAtt to the JSON object passed through `value` and resolved during
   * synthesis.
   *
   * Normally there is no need to use this property since `CfnJson` is an
   * IResolvable, so it can be simply used as a value.
   */
  private readonly value: Reference;

  private readonly jsonString: string;

  constructor(scope: Construct, id: string, props: CfnJsonProps) {
    super(scope, id);

    this.creationStack = captureStackTrace();

    const resourceType = 'Custom::AWSCDKCfnJson';
    const stack = Stack.of(this);

    const provider = CustomResourceProvider.getOrCreate(this, resourceType, {
      runtime: CustomResourceProviderRuntime.NODEJS_12,
      codeDirectory: `${__dirname}/cfn-json-provider`,
    });

    this.jsonString = stack.toJsonString(props.value);

    const resource = new CustomResource(this, 'Resource', {
      serviceToken: provider,
      resourceType,
      properties: {
        Value: this.jsonString,
      },
    });

    this.value = resource.getAtt('Value');
  }

  /**
   * This is required in case someone JSON.stringifys an object which refrences
   * this object. Otherwise, we'll get a cyclic JSON reference.
   */
  public toJSON() {
    return this.jsonString;
  }

  public resolve(_: IResolveContext): any {
    return this.value;
  }
}
