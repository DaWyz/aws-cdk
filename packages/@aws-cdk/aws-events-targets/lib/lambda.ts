import * as events from '@aws-cdk/aws-events';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as sqs from '@aws-cdk/aws-sqs';
import { Names, Stack } from '@aws-cdk/core';
import { addLambdaPermission } from './util';

/**
 * Customize the Lambda Event Target
 */
export interface LambdaFunctionProps {
  /**
   * The event to send to the Lambda
   *
   * This will be the payload sent to the Lambda Function.
   *
   * @default the entire EventBridge event
   */
  readonly event?: events.RuleTargetInput;

  /**
   * The SQS Queue to be used as deadLetterQueue.
   *
   * @default none
   */
  readonly deadLetterQueue?: sqs.IQueue;
}

/**
 * Use an AWS Lambda function as an event rule target.
 */
export class LambdaFunction implements events.IRuleTarget {
  constructor(private readonly handler: lambda.IFunction, private readonly props: LambdaFunctionProps = {}) {

  }

  /**
   * Returns a RuleTarget that can be used to trigger this Lambda as a
   * result from an EventBridge event.
   */
  public bind(rule: events.IRule, _id?: string): events.RuleTargetConfig {
    // Allow handler to be called from rule
    addLambdaPermission(rule, this.handler);

    if (this.props.deadLetterQueue) {
      addToDeadLetterQueueResourcePolicy(rule, this.props.deadLetterQueue);
    }

    return {
      id: '',
      arn: this.handler.functionArn,
      deadLetterConfig: { arn: this.props.deadLetterQueue?.queueArn },
      input: this.props.event,
      targetResource: this.handler,
    };
  }
}


function addToDeadLetterQueueResourcePolicy(rule: events.IRule, queue: sqs.IQueue) {
  const ruleParsedStack = Stack.of(rule);
  const queueParsedStack = Stack.of(queue);

  if (ruleParsedStack.region !== queueParsedStack.region) {
    throw new Error(`Cannot assign Dead Letter Queue to the rule ${rule}. Both the queue and the rule must be in the same region`);
  }

  // Skip Resource Policy creation if the Queue is not in the same account.
  // There is no way to add a target onto an imported rule, so we can assume we will run the following code only
  // where the rule is created.

  if (ruleParsedStack.account === queueParsedStack.account) {
    const policyStatementId = `AllowEventRule${Names.nodeUniqueId(rule.node)}`;

    queue.addToResourcePolicy(new iam.PolicyStatement({
      sid: policyStatementId,
      principals: [new iam.ServicePrincipal('events.amazonaws.com')],
      effect: iam.Effect.ALLOW,
      actions: ['sqs:SendMessage'],
      resources: [queue.queueArn],
      conditions: {
        ArnEquals: {
          'aws:SourceArn': rule.ruleArn,
        },
      },
    }));
  } else {
    // Maybe we could post a warning telling the user to create the permission in the target accout ?
  }
}
