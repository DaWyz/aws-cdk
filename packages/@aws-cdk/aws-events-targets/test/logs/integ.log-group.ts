import * as events from '@aws-cdk/aws-events';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';
import * as targets from '../../lib';

const app = new cdk.App();

const stack = new cdk.Stack(app, 'log-group-events');

const logGroup = new logs.LogGroup(stack, 'log-group', {
  logGroupName: 'MyLogGroupName',
  removalPolicy: cdk.RemovalPolicy.DESTROY,
});

const logGroup2 = new logs.LogGroup(stack, 'log-group2', {
  logGroupName: 'MyLogGroupName2',
  removalPolicy: cdk.RemovalPolicy.DESTROY,
});

const timer = new events.Rule(stack, 'Timer', {
  schedule: events.Schedule.rate(cdk.Duration.minutes(1)),
});
timer.addTarget(new targets.LogGroup(logGroup));

const timer2 = new events.Rule(stack, 'Timer2', {
  schedule: events.Schedule.rate(cdk.Duration.minutes(2)),
});
timer2.addTarget(new targets.LogGroup(logGroup2, {
  event: events.RuleTargetInput.fromObject({
    data: events.EventField.fromPath('$.detail-type'),
  }),
}));

app.synth();

