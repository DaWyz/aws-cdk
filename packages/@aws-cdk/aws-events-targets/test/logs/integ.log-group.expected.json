{
  "Resources": {
    "loggroupB02AAEB1": {
      "Type": "AWS::Logs::LogGroup",
      "Properties": {
        "RetentionInDays": 731
      },
      "UpdateReplacePolicy": "Retain",
      "DeletionPolicy": "Retain"
    },
    "TimerBF6F831F": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "ScheduleExpression": "rate(1 minute)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "loggroupB02AAEB1",
                "Arn"
              ]
            },
            "Id": "Target0"
          }
        ]
      }
    },
    "Timer2B6F162E9": {
      "Type": "AWS::Events::Rule",
      "Properties": {
        "ScheduleExpression": "rate(2 minutes)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "loggroupB02AAEB1",
                "Arn"
              ]
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "f1": "$"
              },
              "InputTemplate": "{\"data\":<f1>}"
            }
          }
        ]
      }
    }
  }
}